File: iosApp/iosApp/Services/ApiClient.swift
import Foundation

// MARK: - Change this to your API base URL
let API_BASE_URL = "http://localhost:8080/api"

enum ApiError: Error, LocalizedError {
    case http(Int, String)
    case network(Error)
    case decode(String)

    var errorDescription: String? {
        switch self {
        case .http(_, let msg): return msg
        case .network(let err): return err.localizedDescription
        case .decode(let msg): return msg
        }
    }
}

final class ApiClient {
    static let shared = ApiClient()
    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.httpAdditionalHeaders = ["Accept-Language": ""]
        return URLSession(configuration: config)
    }()
    private let decoder = JSONDecoder()
    private init() {}

    // MARK: - Auth

    func socialLogin(provider: String, token: String) async throws -> AuthResponse {
        let body: [String: Any] = ["provider": provider, "token": token]
        let data = try await post(path: "/auth/social", body: body, auth: false)
        return try AuthResponse.from(data)
    }

    func refreshTokens() async throws -> TokenPair {
        let body: [String: Any] = ["refresh_token": TokenManager.shared.refreshToken ?? ""]
        let data = try await post(path: "/auth/refresh", body: body, auth: false)
        return try TokenPair.from(data)
    }

    func logout() async throws {
        let body: [String: Any] = ["refresh_token": TokenManager.shared.refreshToken ?? ""]
        _ = try await post(path: "/auth/logout", body: body)
    }

    func getMe() async throws -> ApiUser {
        let data = try await get(path: "/auth/me")
        return ApiUser.from(data)
    }

    // MARK: - Feed

    func getFeed(lat: Double, lng: Double, limit: Int = 20, cursor: String? = nil) async throws -> FeedResponse {
        var params = "lat=\(lat)&lng=\(lng)&limit=\(limit)"
        if let c = cursor { params += "&c=\(c)" }
        let data = try await get(path: "/feed?\(params)")
        return try FeedResponse.from(data)
    }

    // MARK: - Post

    func getMyPosts() async throws -> [ApiFeedPost] {
        let data = try await get(path: "/posts/me")
        guard let arr = data["posts"] as? [[String: Any]] else { return [] }
        return arr.compactMap { ApiFeedPost.from($0) }
    }

    func createPost(text: String, lat: Double, lng: Double) async throws -> [String: Any] {
        let body: [String: Any] = ["post": text, "lat": lat, "lng": lng]
        return try await post(path: "/post", body: body)
    }

    // MARK: - Comments

    func getComments(postId: String) async throws -> [ApiComment] {
        let data = try await get(path: "/post/\(postId)/comments")
        if let arr = data["comments"] as? [[String: Any]] ?? data["data"] as? [[String: Any]] {
            return arr.compactMap { ApiComment.from($0) }
        }
        return []
    }

    func addComment(postId: String, text: String, authorName: String, parentId: String? = nil) async throws -> ApiComment {
        var body: [String: Any] = ["text": text, "author_name": authorName]
        if let pid = parentId { body["parent_id"] = pid }
        let data = try await post(path: "/post/\(postId)/comment", body: body)
        guard let comment = ApiComment.from(data) else { throw ApiError.decode("Invalid comment") }
        return comment
    }

    func voteComment(commentId: String, voteType: String) async throws {
        _ = try await post(path: "/comment/\(commentId)/vote/\(voteType)", body: [:])
    }

    func deleteComment(commentId: String) async throws {
        _ = try await request(method: "DELETE", path: "/comment/\(commentId)", body: nil)
    }

    // MARK: - Signals

    func getSignalKeys(type: String? = nil) async throws -> [SignalKey] {
        let params = type != nil ? "?type=\(type!)" : ""
        let data = try await get(path: "/signal/keys\(params)")
        guard let arr = data["signals"] as? [[String: Any]] else { return [] }
        return arr.compactMap { SignalKey.from($0) }
    }

    func getPostSignals(postId: String) async throws -> PostSignals {
        let data = try await get(path: "/post/\(postId)/signals")
        return PostSignals.from(data)
    }

    func addSignal(postId: String, signalKey: String) async throws {
        _ = try await post(path: "/post/\(postId)/signal", body: ["signal_key": signalKey])
    }

    func removeSignal(postId: String, signalKey: String) async throws {
        _ = try await request(method: "DELETE", path: "/post/\(postId)/signal", body: ["signal_key": signalKey])
    }

    // MARK: - Reports

    func getReportReasons() async throws -> [ReportReason] {
        let data = try await get(path: "/report/reasons")
        guard let arr = data["reasons"] as? [[String: Any]] else { return [] }
        return arr.compactMap { ReportReason.from($0) }
    }

    func reportPost(postId: String, reason: String, detail: String? = nil) async throws -> [String: Any] {
        var body: [String: Any] = ["reason": reason]
        if let d = detail { body["detail"] = d }
        return try await post(path: "/post/\(postId)/report", body: body)
    }

    // MARK: - Internal

    private func get(path: String, auth: Bool = true) async throws -> [String: Any] {
        return try await request(method: "GET", path: path, body: nil, auth: auth)
    }

    private func post(path: String, body: [String: Any], auth: Bool = true) async throws -> [String: Any] {
        return try await request(method: "POST", path: path, body: body, auth: auth)
    }

    private func request(method: String, path: String, body: [String: Any]?, auth: Bool = true) async throws -> [String: Any] {
        guard let url = URL(string: "\(API_BASE_URL)\(path)") else {
            throw ApiError.decode("Invalid URL")
        }

        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if auth, let token = TokenManager.shared.accessToken {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        #if DEBUG
        req.setValue("dev-user-001", forHTTPHeaderField: "X-Dev-User-ID")
        #endif

        if let body = body, method != "GET" {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        #if DEBUG
        print("📡 [\(method)] \(url.absoluteString)")
        if let body = body, method != "GET" {
            print("   Body: \(body)")
        }
        #endif

        let (data, response) = try await session.data(for: req)

        guard let http = response as? HTTPURLResponse else {
            throw ApiError.decode("Invalid response")
        }

        #if DEBUG
        print("   ← \(http.statusCode) (\(data.count) bytes)")
        if let responseStr = String(data: data, encoding: .utf8) {
            print("   Response: \(responseStr)")
        }
        #endif

        // Handle 401 with token refresh
        if http.statusCode == 401 && auth && !path.contains("/auth/refresh") {
            if let _ = TokenManager.shared.refreshToken {
                do {
                    let tokens = try await refreshTokens()
                    TokenManager.shared.saveAuth(accessToken: tokens.accessToken, refreshToken: tokens.refreshToken)
                    return try await request(method: method, path: path, body: body, auth: auth)
                } catch {
                    TokenManager.shared.clear()
                    throw ApiError.http(401, "Session expired")
                }
            }
        }

        if http.statusCode == 204 { return [:] }

        guard (200...299).contains(http.statusCode) else {
            let msg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"] as? String
                ?? String(data: data, encoding: .utf8) ?? "HTTP \(http.statusCode)"
            throw ApiError.http(http.statusCode, msg)
        }

        if data.isEmpty { return [:] }

        let json = try JSONSerialization.jsonObject(with: data)
        if let dict = json as? [String: Any] { return dict }
        if let arr = json as? [[String: Any]] { return ["data": arr] }
        return [:]
    }
}

File: androidApp/src/main/kotlin/com/example/app/android/network/ApiClient.kt
package com.example.app.android.network

import com.example.app.android.network.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Authenticator
import okhttp3.Interceptor
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.Route
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

object ApiClient {

    // ── Change this to your API base URL ──
    const val BASE_URL = "http://10.0.2.2:8080/api"

    private val JSON_MEDIA = "application/json; charset=utf-8".toMediaType()

    private val client: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(15, TimeUnit.SECONDS)
            .addInterceptor(AuthInterceptor())
            .authenticator(TokenRefreshAuthenticator())
            .build()
    }

    // ── Auth ──

    suspend fun socialLogin(provider: String, token: String): Result<AuthResponse> = post(
        "/auth/social",
        JSONObject().put("provider", provider).put("token", token)
    ) { AuthResponse.fromJson(it) }

    suspend fun refreshTokens(): Result<TokenPair> = post(
        "/auth/refresh",
        JSONObject().put("refresh_token", TokenManager.refreshToken)
    ) { TokenPair.fromJson(it) }

    suspend fun logout(): Result<Unit> = post(
        "/auth/logout",
        JSONObject().put("refresh_token", TokenManager.refreshToken)
    ) { }

    suspend fun getMe(): Result<User> = get("/auth/me") { User.fromJson(it) }

    suspend fun updateMe(fields: Map<String, String?>): Result<User> {
        val body = JSONObject()
        fields.forEach { (k, v) -> body.put(k, v ?: JSONObject.NULL) }
        return put("/auth/me", body) { User.fromJson(it) }
    }

    // ── Feed ──

    suspend fun getFeed(
        lat: Double,
        lng: Double,
        limit: Int = 20,
        cursor: String? = null
    ): Result<FeedResponse> {
        val params = buildString {
            append("lat=$lat&lng=$lng&limit=$limit")
            if (cursor != null) append("&c=$cursor")
        }
        return get("/feed?$params") { FeedResponse.fromJson(it) }
    }

    // ── Post ──

    suspend fun createPost(text: String, lat: Double, lng: Double): Result<CreatedPost> = post(
        "/post",
        JSONObject().put("post", text).put("lat", lat).put("lng", lng)
    ) { CreatedPost.fromJson(it) }

    suspend fun getMyPosts(): Result<List<FeedPost>> = get("/posts/me") { json ->
        val arr = json.optJSONArray("posts") ?: JSONArray()
        (0 until arr.length()).map { FeedPost.fromJson(arr.getJSONObject(it)) }
    }

    // ── Comments ──

    suspend fun getComments(postId: String): Result<List<Comment>> = get(
        "/post/$postId/comments"
    ) { json ->
        val arr = json.optJSONArray("comments") ?: json.optJSONArray("data") ?: run {
            // Response might be a raw array - try parsing the body directly
            Comment.listFromJson(JSONArray(json.toString()))
        }
        if (arr is JSONArray) Comment.listFromJson(arr) else emptyList()
    }

    suspend fun addComment(
        postId: String,
        text: String,
        authorName: String,
        parentId: String? = null
    ): Result<Comment> {
        val body = JSONObject()
            .put("text", text)
            .put("author_name", authorName)
        if (parentId != null) body.put("parent_id", parentId)
        return post("/post/$postId/comment", body) { Comment.fromJson(it) }
    }

    suspend fun voteComment(commentId: String, voteType: String): Result<Unit> = post(
        "/comment/$commentId/vote/$voteType", JSONObject()
    ) { }

    suspend fun deleteComment(commentId: String): Result<Unit> = delete("/comment/$commentId")

    // ── Signals ──

    suspend fun getAllSignalKeys(): Result<Map<String, List<SignalKey>>> {
        val lang = java.util.Locale.getDefault().toLanguageTag()
        return request("GET", "/signal/keys/all", null, { json ->
            val result = mutableMapOf<String, List<SignalKey>>()
            json.keys().forEach { typeKey ->
                val arr = json.optJSONArray(typeKey) ?: return@forEach
                result[typeKey] = (0 until arr.length()).map { SignalKey.fromJson(arr.getJSONObject(it)) }
            }
            result
        }, extraHeaders = mapOf("Accept-Language" to lang))
    }

    suspend fun getSignalKeys(type: String? = null): Result<List<SignalKey>> {
        val params = if (type != null) "?type=$type" else ""
        return get("/signal/keys$params") { json ->
            val arr = json.optJSONArray("signals") ?: JSONArray()
            (0 until arr.length()).map { SignalKey.fromJson(arr.getJSONObject(it)) }
        }
    }

    suspend fun getPostSignals(postId: String): Result<PostSignals> = get(
        "/post/$postId/signals"
    ) { PostSignals.fromJson(it) }

    suspend fun syncSignals(postId: String, signalKeys: List<String>): Result<SyncSignalsResponse> {
        val body = JSONObject()
        body.put("signal_keys", JSONArray(signalKeys))
        return put("/post/$postId/signals", body) { SyncSignalsResponse.fromJson(it) }
    }

    // ── Reports ──

    suspend fun getReportReasons(): Result<List<ReportReason>> = get(
        "/report/reasons"
    ) { json ->
        val arr = json.optJSONArray("reasons") ?: JSONArray()
        (0 until arr.length()).map { ReportReason.fromJson(arr.getJSONObject(it)) }
    }

    suspend fun reportPost(
        postId: String,
        reason: String,
        detail: String? = null
    ): Result<ReportResult> {
        val body = JSONObject().put("reason", reason)
        if (detail != null) body.put("detail", detail)
        return post("/post/$postId/report", body) { ReportResult.fromJson(it) }
    }

    // ── Internal helpers ──

    private suspend fun <T> get(
        path: String,
        parse: (JSONObject) -> T
    ): Result<T> = request("GET", path, null, parse)

    private suspend fun <T> post(
        path: String,
        body: JSONObject,
        parse: (JSONObject) -> T
    ): Result<T> = request("POST", path, body, parse)

    private suspend fun <T> put(
        path: String,
        body: JSONObject,
        parse: (JSONObject) -> T
    ): Result<T> = request("PUT", path, body, parse)

    private suspend fun delete(path: String): Result<Unit> =
        request("DELETE", path, null, { })

    private suspend fun <T> request(
        method: String,
        path: String,
        body: JSONObject?,
        parse: (JSONObject) -> T,
        extraHeaders: Map<String, String>? = null
    ): Result<T> = withContext(Dispatchers.IO) {
        try {
            val url = "$BASE_URL$path"
            val requestBody = body?.toString()?.toRequestBody(JSON_MEDIA)

            val request = Request.Builder().url(url).apply {
                when (method) {
                    "GET" -> get()
                    "POST" -> post(requestBody ?: "".toRequestBody(JSON_MEDIA))
                    "PUT" -> put(requestBody ?: "".toRequestBody(JSON_MEDIA))
                    "DELETE" -> if (requestBody != null) delete(requestBody) else delete()
                }
                extraHeaders?.forEach { (key, value) -> header(key, value) }
            }.build()

            val response = client.newCall(request).execute()
            val responseBody = response.body?.string()

            android.util.Log.d("ApiClient", "[$method] $url → ${response.code}")
            android.util.Log.d("ApiClient", "Response: $responseBody")

            if (!response.isSuccessful) {
                val errorMsg = responseBody?.let {
                    try { JSONObject(it).optString("error", it) } catch (_: Exception) { it }
                } ?: "HTTP ${response.code}"
                return@withContext Result.failure(ApiException(response.code, errorMsg))
            }

            if (response.code == 204 || responseBody.isNullOrBlank()) {
                @Suppress("UNCHECKED_CAST")
                return@withContext Result.success(Unit as T)
            }

            val json = try {
                JSONObject(responseBody)
            } catch (_: Exception) {
                // If response is a JSON array, wrap it
                JSONObject().put("data", JSONArray(responseBody))
            }

            Result.success(parse(json))
        } catch (e: ApiException) {
            Result.failure(e)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

class ApiException(val code: Int, message: String) : Exception(message)

private class AuthInterceptor : Interceptor {
    override fun intercept(chain: Interceptor.Chain): Response {
        val original = chain.request()
        val builder = original.newBuilder()

        TokenManager.accessToken?.let {
            builder.header("Authorization", "Bearer $it")
        }

        if (com.example.app.android.DevConfig.BYPASS_API_AUTH) {
            builder.header(
                com.example.app.android.DevConfig.DEV_BYPASS_HEADER,
                com.example.app.android.DevConfig.DEV_BYPASS_VALUE
            )
        }

        return chain.proceed(builder.build())
    }
}

private class TokenRefreshAuthenticator : Authenticator {
    override fun authenticate(route: Route?, response: Response): Request? {
        if (response.code != 401) return null

        val refreshToken = TokenManager.refreshToken ?: return null

        // Avoid infinite loops
        if (response.request.url.encodedPath.contains("/auth/refresh")) return null

        synchronized(this) {
            // Try refreshing
            val refreshBody = JSONObject()
                .put("refresh_token", refreshToken)
                .toString()
                .toRequestBody("application/json".toMediaType())

            val refreshRequest = Request.Builder()
                .url("${ApiClient.BASE_URL}/auth/refresh")
                .post(refreshBody)
                .build()

            val client = OkHttpClient()
            val refreshResponse = client.newCall(refreshRequest).execute()
            val body = refreshResponse.body?.string()

            if (refreshResponse.isSuccessful && body != null) {
                val tokens = TokenPair.fromJson(JSONObject(body))
                TokenManager.saveAuth(tokens.accessToken, tokens.refreshToken)

                return response.request.newBuilder()
                    .header("Authorization", "Bearer ${tokens.accessToken}")
                    .build()
            }

            // Refresh failed — clear tokens
            TokenManager.clear()
            return null
        }
    }
}
